### PR TITLE
CREATE DATABASE two phase commit safe

### PR DESCRIPTION
**Pull request by weinan003**
When a failed raised by `CREATE DATABASE`, relevant files, directories
should not be left over.

Co-authored-by: Asim R P <apraveen@pivotal.io>

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/test/regress/expected/createdb.out [hunk 1]
```diff
@@ -0,0 +1,63 @@
+-- start_ignore
+create language plpythonu;
+-- end_ignore
+CREATE OR REPLACE FUNCTION db_dirs(dboid oid) RETURNS setof text
+  STRICT STABLE LANGUAGE plpythonu
+as $$
+import os
+p = os.popen("find $MASTER_DATA_DIRECTORY/../.. -name %d -type d" % dboid)
+return p.readlines()
+$$;
```
Comment (BaiShaoqi) [inline src/test/regress/expected/createdb.out]: PG funciton `select pg_ls_dir('base/dboid');`  can be used instead ?

Comment (weinan003) [reply src/test/regress/expected/createdb.out]: > PG funciton `select pg_ls_dir('base/dboid');` can be used instead ?
> if so, then `create language plpythonu`; is not needed too.

Well, this UDF is not satisfied. the case needs to check all nodes file path including the master, segments and mirrors.

File: src/backend/commands/dbcommands.c [hunk 2]
```diff
@@ -642,13 +642,21 @@ createdb(const CreatedbStmt *stmt)
 
 			dstpath = GetDatabasePath(dboid, dsttablespace);
 
+			/*
+			 * Register the database directory to PendingDBDelete link list
+			 * for cleanup in txn abort.
+			 */
+			ScheduleDbDirDelete(dboid, dsttablespace, false);
+
```
Comment (BaiShaoqi) [inline src/backend/commands/dbcommands.c]: I may misunderstood here, but is there a bug or case that can illustrate why this was needed excpet the fault injecter case you added ?
because create database cannot run inside a transaction block.

File: src/test/regress/expected/createdb.out [hunk 3]
```diff
@@ -0,0 +1,63 @@
+-- start_ignore
```
Comment (berlin-ab) [inline src/test/regress/expected/createdb.out]: You might find that it is necessary to force the mirrors to create a restart point when replaying a checkpoint record. We found that the suite was restarting mirrors after our test completed and was replaying records that we didn't intend to be replayed.

File: src/test/regress/expected/createdb.out [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+-- should fail
+create database db_with_leftover_files;
+ERROR:  fault triggered, fault name:'create_db_after_file_copy' fault type:'error'  (seg0 127.0.0.1:25432 pid=4555)
+-- Wait until replay_location = flush_location.
+do $$
+declare
+  behind_segs int;
+  i int;
+begin
+  for i in 1..120 loop
```
Comment (berlin-ab) [inline src/test/regress/expected/createdb.out]: We'd made the assumption that the mirrors will eventually catch up (unless down), and therefore did not limit the loop to N retries.

File: src/test/regress/sql/createdb.sql [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+-- inject fault on content0 primary to error out after copying
+-- template db directory
+select gp_inject_fault2('create_db_after_file_copy', 'error', dbid, hostname, port)
+from gp_segment_configuration where content=0 and role='p';
+
+-- should fail
+create database db_with_leftover_files;
+
+-- Wait until replay_location = flush_location.
```
Comment (soumyadeep2007) [inline src/test/regress/sql/createdb.sql]: We could use the XLOG redo no-op mechanism for waiting for mirrors. https://github.com/greenplum-db/gpdb/pull/7792/files#diff-8bc9c09e872d5d30dc0b796ba15a2b3bR33

File: src/test/regress/expected/createdb.out [hunk 6]
```diff
@@ -0,0 +1,63 @@
+-- start_ignore
+create language plpythonu;
+-- end_ignore
+CREATE OR REPLACE FUNCTION db_dirs(dboid oid) RETURNS setof text
+  STRICT STABLE LANGUAGE plpythonu
+as $$
+import os
+p = os.popen("find $MASTER_DATA_DIRECTORY/../.. -name %d -type d" % dboid)
```
Comment (berlin-ab) [inline src/test/regress/expected/createdb.out]: While designing tests for `CREATE TABLESPACE`, @d recommended not depending on `$MASTER_DATA_DIRECTORY`. Also, we chose to use pure python instead of shelling out to `find`.

File: src/test/regress/sql/createdb.sql [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    if behind_segs = 0 then
+      return;
+    end if;
+    perform pg_sleep(0.5);
+  end loop;
+  raise exception 'mirrors took longer than expected to catch up';
+end;
+$$;
+
+set gp_select_invisible=on;
```
Comment (berlin-ab) [inline src/test/regress/sql/createdb.sql]: why is `gp_select_invisible` significant here?

File: src/test/regress/sql/createdb.sql [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+select count(*)=0 as result from
+  (select db_dirs(oid) from pg_database where datname = 'db_with_leftover_files') as foo;
+
+-- start_ignore
```
Comment (berlin-ab) [inline src/test/regress/sql/createdb.sql]: Is this only for debugging purposes?

**Comment by berlin-ab** [comment]
We should consider adding more test scenarios for different phases of two-phase commit, and also some with `panic` instead of `error`.

Note: when panic-ing, consider disabling FTS for the duration of the test.

File: src/test/regress/sql/createdb.sql (hunk 5 above)
Comment (weinan003) [reply src/test/regress/sql/createdb.sql]: wow, that's really cool. thanks a lots

Comment (soumyadeep2007) [reply src/test/regress/sql/createdb.sql]: Welcome!

File: src/test/regress/input/createdb.source [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+CREATE OR REPLACE FUNCTION force_mirrors_to_catch_up() RETURNS VOID AS $$
+BEGIN
+    PERFORM gp_inject_fault2('after_xlog_redo_noop', 'sleep', dbid, hostname, port) FROM gp_segment_configuration WHERE role='m';
+    PERFORM insert_noop_xlog_record();
+    PERFORM gp_wait_until_triggered_fault2('after_xlog_redo_noop', 1, dbid, hostname, port) FROM gp_segment_configuration WHERE role='m';
+    PERFORM gp_inject_fault2('after_xlog_redo_noop', 'reset', dbid, hostname, port) FROM gp_segment_configuration WHERE role='m';
+END;
+$$ LANGUAGE plpgsql;
+
+--
```
Comment (soumyadeep2007) [inline src/test/regress/input/createdb.source]: Can we have a test for a happy path scenario? a successful create without any injected faults

File: src/test/regress/input/createdb.source [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+
+
+--
+--CASE 4: panic after XLOG_XACT_PREPARE on segment
+--
+select gp_inject_fault2('all', 'reset', dbid, hostname, port) from gp_segment_configuration;
+select gp_inject_fault2('end_prepare_two_phase', 'panic', dbid, hostname, port)
+from gp_segment_configuration where content=0 and role='p';
+-- should fail
```
Comment (soumyadeep2007) [inline src/test/regress/input/createdb.source]: I think we should be skipping FTS like here: https://github.com/greenplum-db/gpdb/pull/8128/files#diff-17606b6bbb6095cd09c6e613cf44ae19R125

For context: https://github.com/greenplum-db/gpdb/pull/8118

File: src/test/regress/input/createdb.source [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+    PERFORM gp_inject_fault2('after_xlog_redo_noop', 'reset', dbid, hostname, port) FROM gp_segment_configuration WHERE role='m';
+END;
+$$ LANGUAGE plpgsql;
+
+
+--
+--CASE 0: createdb do well
+--
+create database dowell;
+select count(*)=0 as result from
```
Comment (soumyadeep2007) [inline src/test/regress/input/createdb.source]: Similar to the other scenarios here, we should force_mirrors_to_catch_up before we do the select in order to ensure filesystem changes have been applied on all dbids.

File: src/test/regress/input/createdb.source [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+select gp_inject_fault2('end_prepare_two_phase', 'panic', dbid, hostname, port)
+from gp_segment_configuration where content=0 and role='p';
+select gp_inject_fault2('fts_probe', 'skip', dbid, hostname, port)
+from gp_segment_configuration where content=0 and role='p';
+-- should fail
+create database db4;
+
+select force_mirrors_to_catch_up();
+
+set gp_select_invisible=on;
```
Comment (soumyadeep2007) [inline src/test/regress/input/createdb.source]: Why do we need this GUC to be on?

**Comment by soumyadeep2007** [comment]
LGTM! 
Before merging:
1. Please add a comment in the places where you set `gp_select_invisible` to indicate why it is required.
2. Please squash everything into one commit and provide a summary of the  changes in the resulting commit.

File: src/backend/commands/dbcommands.c (hunk 2 above)
Comment (asimrp) [reply src/backend/commands/dbcommands.c]: @BaiShaoqi CREATE DATABASE runs within an implicitly started transaction.  A command, when run outside a `begin; ... end;` block runs within an implicit transaction.

File: src/test/regress/input/createdb.source [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+CREATE OR REPLACE FUNCTION db_dirs(dboid oid) RETURNS setof text
+  STRICT STABLE LANGUAGE plpythonu
+as $$
+import os
+bash_cmd = "find " + os.getcwd() + "/../../ " + "-name %d -type d"
+p = os.popen(bash_cmd % dboid)
+return p.readlines()
+$$;
+
+--this group udf help test case wait mirror cacheup
```
Comment (asimrp) [inline src/test/regress/input/createdb.source]: typo: s/cacheup/catch up/

File: src/test/regress/output/createdb.source [hunk 14]
```diff
@@ -0,0 +1,463 @@
...
+set gp_select_invisible=on;
+select count(*)=0 as result from
+  (select db_dirs(oid) from pg_database where datname = 'db_with_leftover_files') as foo;
+ result 
+--------
+ t
+(1 row)
+
+-- start_ignore
+select db_dirs(oid) from pg_database where datname = 'db_with_leftover_files';
```
Comment (asimrp) [inline src/test/regress/output/createdb.source]: Should we remove the surrounding ignore tag and have this select as the only validation?  That way, we can get rid of the redundant select above on line 113, 114.  This would work mainly because we expect 0 rows on success.

File: src/test/regress/output/createdb.source [hunk 15]
```diff
@@ -0,0 +1,463 @@
...
+
+select gp_inject_fault2('end_prepare_two_phase', 'panic', dbid, hostname, port)
+from gp_segment_configuration where content=0 and role='p';
+ gp_inject_fault2 
+------------------
+ Success:
+(1 row)
+
+select gp_inject_fault2('fts_probe', 'skip', dbid, hostname, port)
+from gp_segment_configuration where content=0 and role='p';
```
Comment (asimrp) [inline src/test/regress/output/createdb.source]: We must wait until `fts_probe` fault is triggered.  Only after that we can be sure that FTS will not probe the primary and possibly mark it down.

File: src/test/regress/output/createdb.source [hunk 16]
```diff
@@ -0,0 +1,463 @@
...
+(8 rows)
+
+select gp_inject_fault2('end_prepare_two_phase', 'panic', dbid, hostname, port)
+from gp_segment_configuration where content=0 and role='p';
+ gp_inject_fault2 
+------------------
+ Success:
+(1 row)
+
+select gp_inject_fault2('fts_probe', 'skip', dbid, hostname, port)
```
Comment (asimrp) [inline src/test/regress/output/createdb.source]: Use `gp_inject_fault_infinite2()` so that FTS will continue to skip the probes until the fault is reset.  Otherwise, FTS will skip probes only for one probe cycle.

File: src/test/regress/output/createdb.source [hunk 17]
```diff
@@ -0,0 +1,463 @@
...
+ Success:
+ Success:
+ Success:
+ Success:
+ Success:
+ Success:
+ Success:
+(8 rows)
+
+select gp_inject_fault2('before_xlog_xact_prepare', 'error', dbid, hostname, port)
```
Comment (asimrp) [inline src/test/regress/output/createdb.source]: This is identical to `CASE 3: error after XLOG_DBASE_CREATE on segment`.  ERROR on QE at any time after writing the DATABASE CREATE WAL record, whether just after writing the WAL record or just before preparing the transaction, is one and the same.

File: src/test/regress/output/createdb.source [hunk 18]
```diff
@@ -0,0 +1,463 @@
...
+
+-- start_ignore
+select db_dirs(oid) from pg_database where datname = 'db5';
+ db_dirs 
+---------
+(0 rows)
+
+-- end_ignore
+set gp_select_invisible=off;
+--CASE 6: panic before XLOG_XACT_PREPARE on segment
```
Comment (asimrp) [inline src/test/regress/output/createdb.source]: What do we intend to test in this scenario?

Is it the replay of WAL on the segment that PANICs?  Or is it handling of the distributed abort that is broadcast by QD after seeing the PANIC on segment?  Both of these scenarios are already covered.  Replay of WAL is already performed by mirrors, we don't need to test it on primaries again.  Abort of the create database transaction is also covered more than once by the ERROR faults in previous scenarios.

Comment (asimrp) [reply src/test/regress/output/createdb.source]: I am rather interested in seeing what happens upon a PANIC in QD before PREPARE.  How can the left over files be cleaned up in that case, because there is no ABORT WAL record to replay on QD as well as QE.

File: src/test/regress/input/createdb.source [hunk 19]
```diff
@@ -0,0 +1,239 @@
+-- start_ignore
+\! gpconfig -c create_restartpoint_on_chpt_record_replay -v on --skipvalidation;
```
Comment (asimrp) [inline src/test/regress/input/createdb.source]: Why is this needed?  Can we please add some explanation as comment?

My guess is a mirror restart after the test causes some problems?  If so, what exactly are those problems, and why shouldn't we be worried about those?

File: src/test/regress/output/createdb.source [hunk 20]
```diff
@@ -0,0 +1,366 @@
...
+
+-- start_ignore
+select db_dirs(oid) from pg_database where datname = 'db5';
+ db_dirs 
+---------
+(0 rows)
+
+-- end_ignore
+set gp_select_invisible=off;
+--CASE 6: error before XLOG_XACT_DISTRIBUTED_COMMIT on master
```
Comment (asimrp) [inline src/test/regress/output/createdb.source]: This is similar to `CASE 2: error after XLOG_DBASE_CREATE on master`. QD goes through the same code path, as far as cleaning up of leftover directories is concerned.  Segments go through ABORT PREPARED code path rather than ABORT (not prepared) as in CASE 2.  Even that code path on segments is already covered elsewhere in this file.  My recommendation is to remove `CASE 6` altogether.
